pool/shard.go

```diff
@@ -178,10 +194,169 @@ func (pool *shard) firstAvailableProxy(r request) *entry {
+	}
+	//inspired by https://cybernetist.com/2019/01/24/random-weighted-draws-in-go/
+	cdf := make([]float64, size)
+	weights := make([]float64, size)
+	const explorationRate = 0.04
+	for idx, e := range pool.Entries {
+		weights[idx] = float64(e.SuccessRate())
+	}
+	floats.CumSum(cdf, weights)
+	val := distuv.UnitUniform.Rand() * cdf[len(cdf)-1]
+	bandi := sort.Search(len(cdf), func(i int) bool { return cdf[i] > val })
+
+	//fmt.Printf("pool %v (%v/%v)\n", &pool, bandi, size)
+	if bandi == size {
+		bandi -= 1
+	}
+	return &pool.Entries[bandi]
+}
+
+func (pool *shard) BanditUCBProxy() *entry {
+
+	//ctx = app.Log.WithStringer(ctx, "proxy", entry.Proxy)
+	size := len(pool.Entries)
+	if size == 0 {
+		return nil
+	}
+	if size == 1 {
+		return &pool.Entries[0]
+	}
+	//inspired by https://cybernetist.com/2019/01/24/random-weighted-draws-in-go/
+	//cdf := make([]float64, size)
+	weights := make([]float64, size)
+	const explorationRate = 0.005
+	bandi := 0
+
+	nt := 1
+	for _, e := range pool.Entries {
+		nt += e.actions
+	}
+	explore := false
+	sucesses := 0
+	maxv := float64(-1)
+	if rand.Float64() < 0.01 {
+		bandi = rand.Intn(size)
+		explore = true
+	} else {
+		for idx, e := range pool.Entries {
+			sq := math.Sqrt(math.Log(float64(nt))/float64(e.actions+1)) * explorationRate
+			qa := float64(e.SuccessRate() / 100)
+			wall := sq + qa //+ rand.Float64()*0.0000001
+			sucesses += e.Succeed
+			weights[idx] = wall
+			if wall > maxv {
+				maxv = wall
+				bandi = idx
+			}
+		}
+	}
+	//cdf := make([]float64, size)
+	//floats.CumSum(cdf, weights)
+	//val := distuv.UnitUniform.Rand() * cdf[len(cdf)-1]
+	//bandi = sort.Search(len(cdf), func(i int) bool { return cdf[i] > val })
+
+	if bandi == size {
+		bandi -= 1
+	}
+	b := pool.Entries[bandi]
+	fmt.Printf("pool %v %v %.3f \t %v \t(%v/%v) \t%.3f  sel: %v of: %v, suc:%v \t suc/nt%.4f\n", &pool, b.Proxy.IP(), b.SuccessRate(), explore, bandi, size, weights[bandi], b.actions, b.Offered, b.Succeed, float64(b.actions)/float64(nt))
+
+	pool.Entries[bandi].actions += 1
+	return &pool.Entries[bandi]
+}
+func (pool *shard) ThompsonBandit() *entry {
+
+	//ctx = app.Log.WithStringer(ctx, "proxy", entry.Proxy)
+	size := len(pool.Entries)
+	if size == 0 {
+		return nil
+	}
+	if size == 1 {
+		return &pool.Entries[0]
+	}
+	//inspired by https://cybernetist.com/2019/01/24/random-weighted-draws-in-go/
+	//cdf := make([]float64, size)
+	weights := make([]float64, size)
+	const explorationRate = 50
+	bandi := 0
+
+	nt := float64(0.1)
+	sucesses := 0
+	for _, e := range pool.Entries {
+		nt += float64(e.SuccessRate())
+
+	}
+	//explore := false
+
+	mv := float64(-1)
+
+	for idx, e := range pool.Entries {
+		//sq := math.Sqrt(math.Log(float64(nt))/float64(e.actions+1)) * explorationRate
+		//qa := float64(e.SuccessRate() / 100)
+		//wall := sq + qa //+ rand.Float64()*0.0000001
+		//sucesses += e.Succeed
+		//if e.ConsiderSkip(r.in.Context()) {
+		//	continue
+		//}
+		e.Offered += 1
+		a := e.Offered
```

Comment:
Hm... I think it's better to be doing it on "running counters", rather than global counters